src/main/java/com/google/devtools/build/lib/remote/RemoteActionFileSystem.java

```diff
@@ -157,6 +162,91 @@ void flush() throws IOException {
     }
   }
 
+  /**
+   * Inject metadata for non-symlink outputs that were materialized as a symlink to a remote
+   * artifact.
+   *
+   * <p>If a non-symlink output is materialized as a symlink, the symlink has "copy" semantics,
+   * i.e., the output metadata is identical to that of the symlink target. For these artifacts, we
+   * inject their metadata instead of collecting it from the filesystem. This is done for two
+   * reasons:
+   *
+   * <ul>
+   *   <li>It avoids implementing filesystem operations for resolving symlinks and (in the case of a
+   *       tree artifact) listing directories, which are especially tricky since the symlink and its
+   *       target may reside on different filesystems;
+   *   <li>It lets us add a special field to the output metadata to tell the input prefetcher that
+   *       the output should be materialized as a symlink to the original location, which avoids
+   *       fetching multiple copies when multiple symlinks to the same artifact are created in the
+   *       same build.
+   *
+   * @returns Whether the metadata was injected.
+   */
+  private boolean maybeInjectMetadataForSymlink(PathFragment linkPath, Artifact output)
+      throws IOException {
+    if (output.isSymlink()) {
+      return false;
+    }
+
+    Path outputTreePath = remoteOutputTree.getPath(linkPath);
+
+    if (!outputTreePath.exists(Symlinks.NOFOLLOW)) {
+      return false;
+    }
+
+    PathFragment targetPath;
+    try {
+      targetPath = outputTreePath.readSymbolicLink();
+    } catch (NotASymlinkException e) {
+      return false;
+    }
+
+    checkState(
+        targetPath.isAbsolute(),
+        "non-symlink artifact materialized as symlink must point to absolute path");
+
+    if (output.isTreeArtifact()) {
+      TreeArtifactValue metadata = getRemoteInputTreeMetadata(targetPath);
+      if (metadata == null) {
+        return false;
+      }
+
+      SpecialArtifact parent = (SpecialArtifact) output;
+      TreeArtifactValue.Builder injectedTree = TreeArtifactValue.newBuilder(parent);
+      // Avoid a double indirection when the target is already materialized as a symlink.
+      injectedTree.setMaterializationExecPath(
+          metadata.getMaterializationExecPath().orElse(targetPath.relativeTo(execRoot)));
+      // TODO: Check directory content on the local fs to support mixed tree.
+      for (Map.Entry<TreeFileArtifact, FileArtifactValue> entry :
+          metadata.getChildValues().entrySet()) {
+        TreeFileArtifact child =
+            TreeFileArtifact.createTreeOutput(parent, entry.getKey().getParentRelativePath());
+        RemoteFileArtifactValue childMetadata = (RemoteFileArtifactValue) entry.getValue();
+        injectedTree.putChild(child, childMetadata);
+      }
+
+      metadataInjector.injectTree(parent, injectedTree.build());
+    } else {
+      RemoteFileArtifactValue metadata = getRemoteInputMetadata(targetPath);
```

Comment:
I don't think we ever create symlinks to outputs in any of the scenarios where a `RemoteActionFileSystem` is currently used, but it doesn't hurt to allow it. Generally speaking, I'd like this class to trend towards a fully functional filesystem, instead of implementing just the bits internal actions happen to require.

In that vein, `getRemoteInputTreeMetadata` should also fall back to `remoteOutputTree`; I think this would matter if, for example, someone were to call `renameTo` on a directory. However, this isn't necessary for symlinks, so I just left a TODO there for now.

I've removed `Input` from both names to make the purpose clear.